rero_ils/modules/documents/dojson/contrib/marc21tojson_kul/model.py

```diff
@@ -0,0 +1,1963 @@
+    if marc21.langs_from_041_h:
+        self['originalLanguage'] = marc21.langs_from_041_h
+    if marc21.admin_meta_data:
+        self['adminMetadata'] = marc21.admin_meta_data
+    # if adminMetadata/descriptionLanguage contains
+    # "Language of cataloging varies", remove it !
+    try:
+        if self['adminMetadata']['descriptionLanguage'] == \
+                'Language of cataloging varies':
+            self['adminMetadata']['descriptionLanguage'] = 'und'
+    except Exception:
+        self['adminMetadata']['descriptionLanguage'] = 'und'
+    main_type = _ISSUANCE_MAIN_TYPE_PER_BIB_LEVEL.get(
+        marc21.bib_level, 'rdami:1001')
+    sub_type = 'NOT_DEFINED'
+    error = False
+    if marc21.bib_level == 'm':
+        if marc21.is_top_level_record:
+            main_type = 'rdami:1002'
+            sub_type = 'set'
+        else:
+            sub_type = 'materialUnit'
+    else:
+        if marc21.bib_level in _ISSUANCE_SUBTYPE_PER_BIB_LEVEL:
+            sub_type = _ISSUANCE_SUBTYPE_PER_BIB_LEVEL[marc21.bib_level]
+        elif marc21.serial_type in _ISSUANCE_SUBTYPE_PER_SERIAL_TYPE:
+            sub_type = _ISSUANCE_SUBTYPE_PER_SERIAL_TYPE[marc21.serial_type]
+    if main_type == 'rdami:1001':
+        if sub_type not in [
+                'article', 'materialUnit', 'privateFile', 'privateSubfile'
+        ]:
+            error = True
+            sub_type = 'materialUnit'
+    elif main_type == 'rdami:1002':
+        if sub_type not in [
+                'set', 'partIndependentTitle', 'partDependantTitle'
+        ]:
+            error = True
+            sub_type = 'set'
+    elif main_type == 'rdami:1003':
+        if sub_type not in [
+            'serialInSerial', 'monographicSeries', 'periodical'
+        ]:
+            error = True
+            sub_type = 'periodical'
+    elif main_type == 'rdami:1004':
+        if sub_type not in ['updatingWebsite', 'updatingLoose-leaf']:
+            error = True
+            sub_type = 'updatingWebsite'
+    if error:
+        error_print('WARNING ISSUANCE:', marc21.bib_id, marc21.rero_id,
+                    main_type, sub_type, marc21.bib_level, marc21.serial_type)
+    self['issuance'] = {'main_type': main_type, 'subtype': sub_type}
+
+    # Add a paragraph to insert a attribute "type" with
+    # main_type and subtype
+    type = [{"main_type": "docmaintype_other"}]
+    if marc21.record_type == 'a':
+        if marc21.bib_level == 'm':
+            type = [{
+                "main_type": "docmaintype_book",
+                "subtype": "docsubtype_other_book"
+            }]
+
+            field_008 = None
+            field_008 = marc21.get_fields(tag='008')
+            # if it's an electronic book
+            if field_008[0]['data'][23] in ('o', 's'):
+                type = [{
+                    "main_type": "docmaintype_book",
+                    "subtype": "docsubtype_e-book"
+                 }]
+        elif marc21.bib_level == 's':
```

Comment:
I would probably is a dict to map a bib_level into a main and sub types.

Reply:
same code than in marc21tojson (see Gianni's code).